### PR TITLE
Created ad token v4 that uses Base64URLEncoding

### PR DESCRIPTION
**Pull request by sunnywu**

### CODE REVIEW

File: lib/base64.h [hunk 1]
```diff
@@ -32,71 +32,112 @@
...
 
+    class Base64 {
+    private:
+        static std::string Encode(const std::vector<uint8_t>& data, const char* sEncodingTable, bool url) {
             size_t in_len = data.size();
             size_t out_len = 4 * ((in_len + 2) / 3);
             std::string ret(out_len, '\0');
             size_t i;
-            char* p = const_cast<char*>(ret.c_str());
+            char *p = const_cast<char *>(ret.c_str());
```
Comment (jon8787) [inline lib/base64.h]: not your code, but this is undefined behavior, we could use .data() if C++17, or &ret[0] if not

File: lib/base64.h [hunk 2]
```diff
@@ -32,71 +32,112 @@
...
             }
 
+            //remove '=' as Base 64 URL doesn't require padding
+            //https://www.rfc-editor.org/rfc/rfc4648#section-5
+            //and '=' is a reserved char in URL spec
+            if (url)
+            {
+                for(int i =0; i < 3; i++)
+                {
+                    if(ret.at(ret.length() - 1) == '=')
```
Comment (jon8787) [inline lib/base64.h]: minor, if last char is not '=', we're going to check it 4 times :)

File: lib/base64.h [hunk 3]
```diff
@@ -32,71 +32,112 @@
...
+            64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64,
+            64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64
+    };
+
+    static constexpr unsigned char base64URLDecodingTable[] = {
+            64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64,
+            64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64,
+            64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 62, 64, 64,
+            52, 53, 54, 55, 56, 57, 58, 59, 60, 61, 64, 64, 64, 64, 64, 64,
+            64,  0,  1,  2,  3,  4,  5,  6,  7,  8,  9, 10, 11, 12, 13, 14,
```
Comment (jon8787) [inline lib/base64.h:81]: probably need some unit tests for these changes, hard to see if there's any bug in this table

File: lib/base64.h [hunk 4]
```diff
@@ -32,71 +32,112 @@
...
-              64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64,
-              64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64,
-              64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64,
-              64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64,
-              64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64,
-              64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64,
-              64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64
-            };
-
+        static void Decode(const std::string& input, const unsigned char* kDecodingTable, std::vector<uint8_t>& out) {
```
Comment (jon8787) [inline lib/base64.h:125]: we can return the out vector by value, it's efficient in c++11

File: lib/base64.h [hunk 5]
```diff
@@ -118,6 +159,41 @@ namespace macaron {
...
+            }
+            return Decode(input, base64DecodingTable, out);
+        }
+
+        static void DecodeBase64URL(const std::string& input, std::vector<uint8_t>& out) {
+            int inputSizeMod4 = input.size() % 4;
+            //it's within spec https://www.rfc-editor.org/rfc/rfc4648#section-5
+            if(inputSizeMod4 != 0)
+            {
+                std::string input2(input);
```
Comment (jon8787) [inline lib/base64.h:182]: given we're copying the string anyway, can consider doing a string replace with _ and -, and then you don't need the base64URL tables

File: lib/uid2encryption.cpp [hunk 6]
```diff
@@ -38,37 +38,51 @@ namespace uid2
 
     DecryptionResult DecryptToken(const std::string& token, const KeyContainer& keys, Timestamp now, IdentityScope identityScope, bool checkValidity)
 	{
+        std::string headerStr = token.substr(0, 4);
```
Comment (jon8787) [inline lib/uid2encryption.cpp:46]: Do we need to check token > 4 bytes?

Comment (sunnywu) [reply lib/uid2encryption.cpp:46]: fixed

File: test/keygen.h [hunk 7]
```diff
@@ -1,22 +0,0 @@
-#pragma once
```
Comment (sunnywu) [inline test/keygen.h:1]: this file renamed to uid2_token_generator.h further below

File: lib/base64.h (hunk 3 above)
Comment (sunnywu) [reply lib/base64.h:81]: added base_64_encoding_test.cpp to verify the values in these decoding/encoding tables

File: lib/base64.h (hunk 4 above)
Comment (sunnywu) [reply lib/base64.h:125]: i'll leave this as is to minimise the changes

File: lib/base64.h (hunk 5 above)
Comment (sunnywu) [reply lib/base64.h:182]: i have adding more unit tests so i think won't need to do this inefficient string replace

Comment (jon8787) [reply lib/base64.h:182]: right, just to note that this string copy usually requires a memory allocation, which can be much slower than a string replace

File: lib/base64.h (hunk 1 above)
Comment (sunnywu) [reply lib/base64.h]: fixed to 
`   char *p = const_cast<char *>(&ret[0]);`

Comment (jon8787) [reply lib/base64.h]: Can we remove the const_cast?

Comment (sunnywu) [reply lib/base64.h]: Removed

**Pull request merged**